Rationalize name-char (#1008)
unicode-org/message-format-wg: spec/syntax.md

> implementations can often substitute checking for actually applying normalization
> to _name_ values.

The _names_ are [immutable identifiers](https://www.unicode.org/reports/tr31/#Immutable_Identifier_Syntax).
They are similar to <cite>Namespaces in XML 1.0</cite>'s [NCName](https://www.w3.org/TR/xml-names/#NT-NCName),

> but have been updated to be more consistent.

Reviewer: Given how far we've departed from that spec, this note seems only relevant within the history of the spec, and not with where we're ending up with this PR. I'd prefer dropping it, and moving the preceding sentence (if keeping) above the preceding note.

Author: ok.

                  / %x10000-1FFFD   ;          omit NChar: %x1FFFE-1FFFF
                  / %x20000-2FFFD   ;          omit NChar: %x2FFFE-2FFFF
                  / %x30000-3FFFD   ;          omit NChar: %x3FFFE-3FFFF
                  / %x40000-4FFFD   ;          omit NChar: %x4FFFE-4FFFF
                  / %x50000-5FFFD   ;          omit NChar: %x5FFFE-5FFFF
                  / %x60000-6FFFD   ;          omit NChar: %x6FFFE-6FFFF
                  / %x70000-7FFFD   ;          omit NChar: %x7FFFE-7FFFF
                  / %x80000-8FFFD   ;          omit NChar: %x8FFFE-8FFFF
                  / %x90000-9FFFD   ;          omit NChar: %x9FFFE-9FFFF
                  / %xA0000-AFFFD   ;          omit NChar: %xAFFFE-AFFFF
                  / %xB0000-BFFFD   ;          omit NChar: %xBFFFE-BFFFF
                  / %xC0000-CFFFD   ;          omit NChar: %xCFFFE-CFFFF
                  / %xD0000-DFFFD   ;          omit NChar: %xDFFFE-DFFFF
                  / %xE0000-EFFFD   ;          omit NChar: %xEFFFE-EFFFF
                  / %xF0000-FFFFD   ;          omit NChar: %xFFFFE-FFFFF
                  / %x100000-10FFFD ;          omit NChar: %x10FFFE-10FFFF

> 
> name-char  = name-start / DIGIT
>                   / %x2D-2E         ; «-.»

Reviewer: ```suggestion
name-char  = name-start / DIGIT / "-" / "."
```

                  / %xC0000-CFFFD ;          omit NChar %xBFFFE-BFFFF
                  / %xD0000-DFFFD ;          omit NChar %xCFFFE-CFFFF
                  / %xE0000-EFFFD ;          omit NChar %xDFFFE-DFFFF,
                                  ;          omit NChar %xEFFFE-EFFFF %xFFFFE-FFFFF %x10FFFE-10FFFF,
                                  ;          omit Co %xF0000-FFFFD %x100000-10FFFD
name-char  = name-start / DIGIT
                  / %x2D-2E       ; 【-.】    omit Cc %x0-1F, Whitespace 【 】, Ascii 【!"#$%&'()*+,】
```

> [!NOTE]
> Syntactically, the definitions of `identifier` and `name-char` provide backwards compatibility over time by allowing a stable,
> wide range of characters.
> So when there is a new character in a version of Unicode, it can be used in any conformant implementation of Message Format.
> The definition currently excludes:
> * Most ASCII except for letters and characters used for numbers
>    * This avoids conflicts with syntax characters, and reserves some characters for future syntax.
> * Bidirectional controls (`Bidi_C`)
> * Control characters (`GC=Cc`, but not Format characters: `GC=Cf`)
> * Whitespace characters (`WSpace`)

> > * Isolated Surrogate characters (`GC=Cs`)

Reviewer 2: We don't allow them in pairs either. A UTF-16 encoded message might have surrogate code units in it, but they are interpreted as the encoded code point. So:

```suggestion
> * Surrogate code points (`GC=Cs`)
```

> * Private use characters (`GC=Co`)
> * Non-Characters (`NChar`)
>

> > Although syntactically a wide range of characters are included,
> > when function and implementations and message authors are creating new identifiers (for functions, options, variables, …),
> > it is strongly recommended that they conform to the following to minimize confusion.
> > These are also recommended for Message Format linter implementations.

Reviewer 2: This is normative language in a "NOTE", which is verboten in our specification. There is no need to make this a note, since this is the specification and is meant to describe what is going on. Something like:

```suggestion
This syntax allows a wide range of characters in _names_ and _identifiers_.
Implementers and authors of _functions_ and _messages_,
including _functions_, _options_, and _operands_ (variable names),
SHOULD avoid creating _names_ that could produce confusion or harm usability
by choosing names consistent with the following guidelines.
MessageFormat tools, such as linters, SHOULD warn when _names_ chosen by users
violate these constraints.
```

(Note that it is `MessageFormat` not `Message Format`)
